ddtrace/contrib/botocore/patch.py

```diff
@@ -481,212 +482,260 @@ def patched_api_call(original_func, instance, args, kwargs):
-                        )
-                        span.name = schematize_cloud_messaging_operation(
-                            trace_operation,
-                            cloud_provider="aws",
-                            cloud_service="kinesis",
-                            direction=SpanDirection.OUTBOUND,
-                        )
-                    if endpoint_name == "sns" and operation == "Publish":
-                        inject_trace_to_sqs_or_sns_message(
-                            params,
-                            span,
-                            endpoint_service=endpoint_name,
-                            pin=pin,
-                            data_streams_enabled=config._data_streams_enabled,
-                        )
-                        span.name = schematize_cloud_messaging_operation(
-                            trace_operation, cloud_provider="aws", cloud_service="sns", direction=SpanDirection.OUTBOUND
-                        )
-                    if endpoint_name == "sns" and operation == "PublishBatch":
-                        inject_trace_to_sqs_or_sns_batch_message(
-                            params,
-                            span,
-                            endpoint_service=endpoint_name,
-                            pin=pin,
-                            data_streams_enabled=config._data_streams_enabled,
-                        )
-                        span.name = schematize_cloud_messaging_operation(
-                            trace_operation, cloud_provider="aws", cloud_service="sns", direction=SpanDirection.OUTBOUND
-                        )
-                except Exception:
-                    log.warning("Unable to inject trace context", exc_info=True)
-
-            if params and not config.botocore["tag_no_params"]:
-                aws._add_api_param_span_tags(span, endpoint_name, params)
+            try:
+                start_ns = time_ns()
+                func_run = True
+                result = original_func(*args, **kwargs)
+            except Exception as e:
+                func_run_err = e
+            if result is not None and "Messages" in result and len(result["Messages"]) >= 1:
+                message_received = True
+                ctx = extract_DD_context(result["Messages"])
+        elif endpoint_name == "kinesis" and operation == "GetRecords":
+            try:
+                start_ns = time_ns()
+                func_run = True
+                result = original_func(*args, **kwargs)
+            except Exception as e:
+                func_run_err = e
+            if result is not None and "Records" in result and len(result["Records"]) >= 1:
+                message_received = True
+                ctx = extract_DD_context(result["Records"])
+
+    # we don't want to create a span for empty poll consumer operations
+    if (func_run and message_received) or config.empty_poll_enabled or not func_run:
```

Comment:
The way this is written is cognitively complex (its not clear what func_run means, for instance, nor how empty_poll_enabled relates to endpoints other than sqs).

I recommend refactoring this by having a "default" trace function that is called for endpoints, then overriding based on the endpoint.

```
if endpoint == sqs:
    special_sqs_trace(*args, **kwargs)
elif endpoint == "events":
    default_trace(*args, **kwargs)
```

Reply:
The idea is we can have special trace logic per endpoint and make it clear what logic is being applied to which endpoint

Reply:
Done!